### ISSUE TITLE
New Rule `duplicate-properties` - allow for property fallbacks

### ISSUE THREAD

**Issue by MadLittleMods** [stylelint/stylelint#404]
`rule-no-duplicate-properties` doesn't handle enough.

I would like to disallow duplicate properties but only if they are using the same units. This would help with defining fallbacks when the value fails.

 `duplicate-properties` would have these possible values:

 - `never`: Throw whenever there are duplicate property names.
 - `no-duplicate-units`: Throw when there are declarations with the same property name and same unit type.

The following would be allowed with `'duplicate-properties': 'no-duplicate-units'`:
```css
.foo {
	color: #aaaa00;
	color: rgba(255, 255, 0, 0.5);
}

.bar {
	height: 500px;
	height: 50vh;
}

.baz {
	height: 500px;
	height: calc(100% - 200px);
}
```

---

[`postcss-value-parser`](https://github.com/TrySound/postcss-value-parser) will probably be pretty useful for this.

**Comment by jeddy3** [comment]
> `rule-no-duplicate-properties` doesn't handle enough.

I agree with that :) The rule does what it says on the tin, but it's not very practical. 

Your suggestion of checking for duplicated units sounds very worthwhile. 

I'd like to get @davidtheclark input on where this functionality should live though before opening this out to a PR. Throwing a couple of suggestions out there:

* We could deprecate `rule-no-duplicate-properties`, in favour of `rule-duplicate-properties: "never"|"no-duplicated-units"`
* We could make `rule-no-duplicate-properties` our first `-no-` rule to accept an optional options object i.e. `rule-no-duplicate-properties: { allow-differing-units: true|false }.

I'm leaning towards the latter as we consistently used `-no-` for rules where a "positive" option doesn't make sense e.g. all rules _must be_ duplicates.

Thoughts?

+1 for `postcss-value-parser` btw :) I've been eyeing it up for some time to possibly help out with the implementation of `property-unit-blacklist`.

**Comment by davidtheclark** [comment]
> We could make rule-no-duplicate-properties our first -no- rule to accept an optional options object i.e. `rule-no-duplicate-properties: { allow-differing-units: true|false }.

@jeddy3: I like this solution. I think it makes good sense.

@MadLittleMods: Also thanks for pointing out postcss-value-parser! I don't think I was aware of it for some reason, and it will definitely come in handy!

**Comment by jeddy3** [comment]
> I like this solution. I think it makes good sense.

Good stuff. We might need to think about the config structure of this e.g. do we pass null in the primary option slot i.e.

```js
rule-no-duplicate-properties: [2, null, { allow-differing-units: true|false }]
```

Or move these optional options into the primary slot, which kind of makes the config structure less consistent e.g.

```js
rule-no-duplicate-properties: [2, { allow-differing-units: true|false }]
```

I'm not sure which is best.

**Comment by davidtheclark** [comment]
I prefer the second (no `null`).

**Comment by jeddy3** [comment]
Coolio. I think the fact that it's an object helps to distinguish it from a primary option. 

e.g. 

```js
"indentation": [2, "tab", { except: ["value"] }],
"rule-no-duplicate-properties": [2, { allow-differing-units: true|false }]
```

It feels pretty clear here that we have a "normal" rule with a primary option and with an optional options object, followed by a `-no-` rule with just an optional options object.

**Comment by davidtheclark** [comment]
Actually this adds a lot of tricky edge cases because there is not a 1:1 declaration:unit relation. What would we do in situations where there are multiple units in a single declaration? E.g.

```css
.foo {
  margin: 2px 3em;
  margin: 2px 2rem;
}

.foo {
  transition: height 0.3s;
  transition: width 200ms;
}
```

Additionally, I don't think postcss-value-parser can handle multiple units right now: https://github.com/TrySound/postcss-value-parser/issues/11

I wonder if a simpler way to approach this problem would be to `ignore` certain properties. For example, if you @MadLittleMods know that you're going to be repeating `font-size` with `px` and `rem` values throughout your stylesheets, you could `ignore: [ 'font-size' ]`. Would that solve the main problem, or do we think that's not enough?

**Comment by jeddy3** [comment]
>  know that you're going to be repeating font-size with px and rem values throughout your stylesheets, you could ignore: [ 'font-size' ]

This sounds the right balance to me.

**Comment by TrySound** [comment]
@davidtheclark It can handle, but unit parser should be applied to word, not to whole value.

**Comment by MadLittleMods** [comment]
@davidtheclark I don't think an `ignore` option would be enough.

For viewport dimensions or color `rgba` for example, I would still want to catch duplicate properties that may appear by accident but still allow the fallback.

---

For multiple values, we would look for a different/unsafe unit type anywhere inside. In the example of `margin: 2px 2rem;`, if a browser didn't support `rem` it would invalidate that whole declaration.

I could see having a whitelist of `unsafe` units that would be a user customized.

```
{
	units: [
		vh',
		'vw',
		'rem'
	],
	functions: [
		'rgba'
	]
	display: [
		'flex',
		'inline-flex'
	],
}
```

**Comment by davidtheclark** [comment]
@TrySound : Ah, ok. Thanks for the clarification. :)

@MadLittleMods : I don't know ... as you have shown, that becomes pretty complicated fast. I'm not eager to build it myself at this point. If you or @jeddy3 want to mock up documentation or tests for what the finished rule would do, we could review it and see if it's reasonable.

**Comment by jeddy3** [comment]
> I'm not eager to build it myself at this point

Same as I'm struggling to see the benefit outweigh the effort. Especially as I mostly use [fallback plugins](https://github.com/postcss/postcss#fallbacks) if I need to convert things like `rgba()` values and `rem` units. `rule-no-duplicate-properties`, in it's current form, is helpful then. 

But like @davidtheclark said, feel free to mock up docs and/or tests if you'd like to keep going with this one @MadLittleMods :)

**Comment by MadLittleMods** [comment]
@jeddy3 @davidtheclark I personally would use this for the viewport unit fallbacks. And I have also had a situation where I have `flex` fallback to `inline-block` but this is probably niche.

For general unit fallback, it could just be another postcss fallback plugin.

 - `width: unit(45vh, 450px);`
 - `background-color: unit(rgba(0, 0, 0, 0.5), #555555);`

**Comment by davidtheclark** [comment]
@MadLittleMods : That does seem like an interesting plugin idea.

**Comment by MadLittleMods** [comment]
I just released [`postcss-fallback`](https://www.npmjs.com/package/postcss-fallback) which should tackle this problem in a different way [as described in a previous comment](https://github.com/stylelint/stylelint/issues/404#issuecomment-148503290).

---

Input:
```css
.foo {
    width: fallback(45vh, 450px);
    foo: fallback(bar, baz, qux, corge);
}
```

Output:
```css
.foo {
    width: 450px;
    width: 45vh;
    foo: corge;
    foo: qux;
    foo: baz;
    foo: bar;
}
```

**Comment by jeddy3** [comment]
Nice plugin :)

Closing this issue.
